### ISSUE TITLE
add a maven mode for using sbt on existing maven projects

### ISSUE THREAD

**Issue by jstrachan** [sbt/sbt#287]
There's a ton of maven projects out there; e.g. pretty much all JVM based stuff at apache & codehaus uses maven for example.

Asking/waiting for folks to rewrite their entire builds in SBT is not gonna happen any time soon. If you want SBT to gain world domination sooner (you may not care :) you could do the following...
- support a 'maven mode' for sbt. Just like the latest SBT can boot up with little/no configuration; if there's no sbt configuration at all but there is a pom.xml, auto-detect it and boot up SBT in 'maven mode'. 
- this would auto-create a little bootstrap sbt configuration file if it doesn't exist already - which turns on maven-mode (and could later on get tweaked)
- maven-mode would just read the pom.xml files using Aether (not Ivy!) for faithful resolving of all the maven metadata, multi-module dependency trees & properties & profiles etc; then use SBT for the actual tasks of compiling & running tests. e.g. see [maven-sbt](https://github.com/rossabaker/maven-sbt)
## Benefits

You instantly get a ton more prospective users. Its now a no brainer to start using SBT as a build tool on mvn projects. You can then hook maven users with your cool SBT incremental compile crack & give them fast compiles on scala.

Over time folks might even migrate away from the mvn build & mvn plugins to SBT. But its zero risk. So folks could keep their mvn builds for things like releasing; but use SBT for day to day stuff. Or keep both builds in parallel for a while until folks build up enough trust in SBT (or when SBT gets to 1.0 or something :)  

Today asking folks to totally replace their maven build is a pretty big ask - particularly if the maven build is already done and works fine mostly. Without maven-mode, there's a huge barrier to adoption for projects which already exist & use maven; most folks have better things to do with their time than re-write build configurations :) However using maven-mode - which would be pretty simple and quick to create I'd have thought - it'd be a no brainer for folks to dip their toes into SBT & they could get hooked :). Bwahahahaha.... :)

**Comment by jstrachan** [comment]
My intention with this issue wasn't so much for SBT to implement maven or host maven plugins (we can use Maven for that :) - its more about reusing all the existing maven metadata (pom.xml files) to bootstrap SBT & to avoid folks having to rewrite all their pom.xml files to be able to reuse SBT.

**Comment by rossabaker** [comment]
GitHub Issues doesn't support votes, so here's my +1.

Not only is porting a Maven build difficult, it's forbidden in some shops.  A team might be allowed to try Scala, but the SCM team that manages dozens of builds is not going to change their process.  I've consulted in places where I might have sold the architect on Scala, but sbt would have been a non-starter.  This idea would let development enjoy sbt's shell and incremental compilation without disrupting the wider IT department.  In the enterprise, Maven is as much a part of interop as the JVM itself.

**Comment by viktorklang** [comment]
If it's doable, it sounds awesome

**Comment by jstrachan** [comment]
Back in the 0.7x days it was pretty easy to get SBT to bootstrap itself dependency wise from maven poms - though there were some gremlins in the SBT+Ivy approximations of maven metadata. 

Using something like maven-sbt (or even just direct Aether) to build a tree of maven metadata would mean SBT doesn't need to worry about it; Aether is pretty easy to wire in and faithfully supports all maven metadata. 

Am not sure how easy it is these days to bootstrap SBT from the Aether generated metadata though; shouldn't have thought its too hard though as its just trees of projects & dependencies (with scopes and whatnot) and you can let Aether do all the downloads & repo handling if required.

**Comment by imikushin** [comment]
+1 It'll be killer.

**Comment by hedefalk** [comment]
+1

**Comment by pokerjocke70** [comment]
+1  that would make the transition much smoother

**Comment by harrah** [comment]
Please see https://github.com/harrah/xsbt/wiki/Build-Loaders.  In particular, see the "Custom Builder" section where it describes the hooks that can be used to implement this and gives reading from a pom as an example.  Although the implementations of readPom and pomSettings were not given in that documentation, I wrote prototype implementations at the time to verify it worked.  It is easy to write the proof of concept, but it is a lot of effort to make real world projects work.

It is a supported use case for someone to be able to do this without modifying sbt, but it is beyond the resources of the core team to make a practical implementation.

**Comment by chirino** [comment]
Yep.  +1

**Comment by yoshinoriishii** [comment]
harrah
Would you show me the prototype for maven-mode ?
I want to try it.

**Comment by viktorklang** [comment]
Classy James ;-)

**Comment by jstrachan** [comment]
Whoah! No idea where that came from! :)

**Comment by taylorleese** [comment]
+1 on maven mode

**Comment by gubbes** [comment]
Great idea. +1

**Comment by ericacm** [comment]
+1 - is this closed because it got folded into another issue, or?

**Comment by crankydillo** [comment]
+1

I have a hard enough time convincing 10 or so devs that we should use Scala on some projects.  Our company has a significant investment in Maven, Jenkins, Sonar, etc., which serves a much wider set of devs.  Not being able to easily utilize that makes my arguments substantially weaker.

I currently use mvn to download all dependencies into ${project}/lib.  My hard drive doesn't like it!
